### PR TITLE
[LegalizeTypes][AMDGPU]: Allow for scalarization of insert_subvector

### PR DESCRIPTION
**Pull request by jrbyrnes**
Legalization for when the inserted subvector is to be scalarized.

https://godbolt.org/z/vx3joWqoh

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/scalarize-insert-subvector.ll [hunk 1]
```diff
@@ -0,0 +1,26 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 < %s | FileCheck -check-prefixes=GCN %s
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/scalarize-insert-subvector.ll:2]:
```suggestion
; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx942 < %s | FileCheck -check-prefix=GCN %s
```

File: llvm/test/CodeGen/AMDGPU/scalarize-insert-subvector.ll [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  %shuffle0 = shufflevector <2 x i32> %load1, <2 x i32> poison, <1 x i32> <i32 1>
+  %bitcast0 = bitcast <1 x i32> %shuffle0 to <2 x half>
+  %bitcast1 = bitcast <2 x i32> %load0 to <4 x half>
+  %shuffle1 = shufflevector <2 x half> %bitcast0, <2 x half> poison, <4 x i32> <i32 0, i32 1, i32 poison, i32 poison>
+  %shuffle2 = shufflevector <4 x half> %bitcast1, <4 x half> %shuffle1, <4 x i32> <i32 0, i32 1, i32 4, i32 5>
+  store <4 x half> %shuffle2, ptr addrspace(3) %outptr
+  ret void
+}
+;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
+; SI: {{.*}}
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/scalarize-insert-subvector.ll:26]:
```suggestion

```

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 3]
```diff
@@ -882,6 +885,16 @@ SDValue DAGTypeLegalizer::ScalarizeVecOp_CONCAT_VECTORS(SDNode *N) {
   return DAG.getBuildVector(N->getValueType(0), SDLoc(N), Ops);
 }
 
+/// The inserted subvector is to be scalarized - use insert vector element
+/// instead.
+SDValue DAGTypeLegalizer::ScalarizeVecOp_INSERT_SUBVECTOR(SDNode *N,
+                                                          unsigned OpNo) {
+  auto Elt = GetScalarizedVector(N->getOperand(OpNo));
+  auto VecOp = N->getOperand(1 - OpNo);
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:893]: No auto? VecOp could be a better name, since both of the operands are typically vectors

**Comment by jrbyrnes** [comment]
The implementation assumes that the scalarized operand is the second input vector operand -- I've made it explicit.

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp [hunk 4]
```diff
@@ -889,10 +889,13 @@ SDValue DAGTypeLegalizer::ScalarizeVecOp_CONCAT_VECTORS(SDNode *N) {
...
 SDValue DAGTypeLegalizer::ScalarizeVecOp_INSERT_SUBVECTOR(SDNode *N,
                                                           unsigned OpNo) {
-  auto Elt = GetScalarizedVector(N->getOperand(OpNo));
-  auto VecOp = N->getOperand(1 - OpNo);
-  return DAG.getNode(ISD::INSERT_VECTOR_ELT, SDLoc(N), VecOp.getValueType(),
-                     VecOp, Elt, N->getOperand(2));
+  // We should not be attempting to scalarize the containing vector
+  assert(OpNo == 1);
+  SDValue Elt = GetScalarizedVector(N->getOperand(OpNo));
+  SDValue ContainingVec = N->getOperand(1 - OpNo);
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:895]: Might as well assume 0 if you're asserting above

**Review by arsenm — APPROVED** [review]

**Comment by jrbyrnes** [comment]
Pushed for web UI

**Pull request merged**